### ISSUE TITLE
Single value model predictions

### ISSUE THREAD

**Issue by JRSlotman** [dhopp1/nowcastLSTM#4]
I've been struggling to get the model working on my own sample data, which is a selection of data from World Bank GEM (see attached). Whatever the parameters I choose for the training data and the model, I only get a single value for the entire series. What am I doing wrong?

[nowcast_GEM.zip](https://github.com/dhopp1/nowcastLSTM/files/6933754/nowcast_GEM.zip)

**Comment by dhopp1** [comment]
It's because you were using levels/actual values instead of growth rates. In the attached code I added the transformations for period on period growth rates, and you see that fixes your one prediction for all periods problem.

[nowcasting_lstm.txt](https://github.com/dhopp1/nowcastLSTM/files/6933995/nowcasting_lstm.txt)

However you still aren't getting very good predictions, partially because the growth rates should probably be seasonally adjusted. Beyond that, then you can play with the hyper parameters, the ones I've found to make the most difference are below, plus some good values to test. You can just do it with nested for loops. Just make sure you're keeping track of the performance (say MAE on the train and test set) and hyper parameters used for that model.

```
n_timesteps=c(6, 12, 24)
train_episodes=c(20, 50, 100, 200)
n_hidden=c(10, 20, 40)
n_layers=c(1,2)
batch_size=c(10, 30, 50)
```

I also noticed you have a lot of heteroskedasticity in the target variable. When you plot the `urx` growth rates you can see the variance increases a lot over time. I didn't check all the independent variables, but you should see if that behavior is duplicated in any of them. Otherwise you should look into adding new variables, as you may not have any that are capturing the later (or earlier) behavior of the target series. Also you should probably drop earlier periods from training where many series are missing (say pre-2000). But you can experiment empirically.

My hunch is you probably will need to look into adding new explanatory variables, usually if I have strong predictive variables I can get the model to really overfit and mirror the train period almost exactly (of course not good in practice, but means that there is enough information in the explanatory variables to result in the target series). Playing around quickly with some LSTM models on your data I couldn't really get that close to overfitting, except in the earlier period. The later it was, the more off the training predictions, more evidence that your explanatory data could be missing something important for predicting the target series later on. I also ran a quick dynamic factor model as a sanity check and got similarly poor predictions.

Let me know how things go or if you need further pointers! You can even try using the information in the example dataset, those are all already seasonally adjusted growth rates of many macroeconomic series. If you end up wanting to use any I can send you the code for getting/updating them yourself.

**Comment by JRSlotman** [comment]
Thank you for this very elaborate response! I'm new to machine learning so I am not yet very confident with playing around with all the parameters, but I am just going to give it a try. It seems that adding growth rates (YOY and QOQ) for most of the explanatory variables can actually get us pretty close to overfitting on the training data, but I agree that I should probably need a few more to compensate for all the missing values and to better explain the recent volatility. I'll give it a try and keep you updated on the progress.

**Comment by dhopp1** [comment]
I looked at your variables a bit closer now, and realized most are already seasonally adjusted, but also that YoY seems to work better. I also realized CPI and URX could remain untransformed, just / 100 to be the same units as the other YoY growth rates. 

Playing around a bit with that model then I could get much better results. This is with the default model parameters, `n_timesteps=4`, `n_models=5`, everything but CPI and URX in YoY growth rates ([nowcasting_lstm.txt](https://github.com/dhopp1/nowcastLSTM/files/6936979/nowcasting_lstm.txt)):
![image](https://user-images.githubusercontent.com/6787728/128311637-19118dd1-f4bd-48b8-b3ce-48b8fead794d.png)

This is the test set, I set training period only until 2017 to see how it performed in more "normal" times in 2018 and 2019. So just predicting directly the unemployment rate. Not perfectly, but it's able to pick up on the crazy signals in 2020 and 2021. I bet if you added in a couple variables more tightly connected with unemployment, e.g. maybe [business confidence index](https://data.oecd.org/leadind/business-confidence-index-bci.htm), and played around more with the hyperparameters you could end up with a pretty good model.
